src/NHibernate/Linq/NestedSelects/NestedSelectRewriter.cs

```diff
@@ -59,98 +59,104 @@ public static void ReWrite(QueryModel queryModel, ISessionFactory sessionFactory
 
 			queryModel.ResultOperators.Add(new ClientSideSelect2(lambda));
-			queryModel.ResultOperators.Add(new ClientSideSelect(Expression.Lambda(resultSelector, @group)));
+			queryModel.ResultOperators.Add(new ClientSideSelect(Expression.Lambda(resultSelector, group)));
 
 			var initializers = elementExpression.Select(e => ConvertToObject(e.Expression));
 
-			queryModel.SelectClause.Selector = Expression.NewArrayInit(typeof (object), initializers);
+			queryModel.SelectClause.Selector = Expression.NewArrayInit(typeof(object), initializers);
 		}
 
-		private static Expression ProcessExpression(QueryModel queryModel, ISessionFactory sessionFactory, Expression expression, List<ExpressionHolder> elementExpression, ParameterExpression @group)
+		private static Expression ProcessExpression(QueryModel queryModel, Expression expression, List<ExpressionHolder> elementExpression,
+			ParameterExpression group, VisitorParameters parameters)
 		{
-			var memberExpression = expression as MemberExpression;
-			if (memberExpression != null)
-				return ProcessMemberExpression(sessionFactory, elementExpression, queryModel, @group, memberExpression);
-			
-			var subQueryExpression = expression as SubQueryExpression;
-			if (subQueryExpression != null)
-				return ProcessSubquery(sessionFactory, elementExpression, queryModel, @group, subQueryExpression.QueryModel);
-			
+			if (expression is MemberExpression memberExpression)
+				return ProcessMemberExpression(elementExpression, queryModel, group, memberExpression, parameters);
+
+			if (expression is SubQueryExpression subQueryExpression)
+				return ProcessSubquery(elementExpression, queryModel, group, subQueryExpression.QueryModel, parameters);
+
 			return null;
 		}
 
-		private static Expression ProcessSubquery(ISessionFactory sessionFactory, ICollection<ExpressionHolder> elementExpression, QueryModel queryModel, Expression @group, QueryModel subQueryModel)
+		private static Expression ProcessSubquery(ICollection<ExpressionHolder> elementExpression, QueryModel queryModel, Expression group,
+			QueryModel subQueryModel, VisitorParameters parameters)
 		{
 			var subQueryMainFromClause = subQueryModel.MainFromClause;
 
-			var restrictions = subQueryModel.BodyClauses
-											.OfType<WhereClause>()
-											.Select(w => new NhWithClause(w.Predicate));
-
-			var join = new NhJoinClause(subQueryMainFromClause.ItemName,
-										subQueryMainFromClause.ItemType,
-										subQueryMainFromClause.FromExpression,
-										restrictions);
+			var restrictions = subQueryModel.BodyClauses.OfType<WhereClause>().ToList();
+			var join = new AdditionalFromClause(
+				subQueryMainFromClause.ItemName,
+				subQueryMainFromClause.ItemType,
+				subQueryMainFromClause.FromExpression);
 
-			queryModel.BodyClauses.Add(@join);
+			parameters.AddLeftJoin(join, restrictions);
+			queryModel.BodyClauses.Add(join);
+			foreach (var withClause in restrictions)
+			{
+				queryModel.BodyClauses.Add(withClause);
```

Comment:
Adding them to the query model allows them to get transformed by any later `TransformExpressions`. This has no other usage currently.

They will not be added as HQL where because the query model visitor now check with visitor parameters each where clause for not being known as a with clause.